### DISCUSSION TITLE
On demand fuzzy find downloading (and perhaps uploading too?)

### DISCUSSION THREAD

**Discussion by kalyanoliveira** [abraunegg/onedrive#2752] [Show and tell]
This discussion follows the two previous questions ([here](https://github.com/abraunegg/onedrive/discussions/2750) and [here](https://github.com/abraunegg/onedrive/discussions/2749)) that I posted in the [Discussions section](https://github.com/abraunegg/onedrive/discussions) of [abraunegg/onedrive](https://github.com/abraunegg/onedrive).

I'm just looking for a specific way of accessing my OneDrive files, really.

Here's the kind of behavior that I'm after: on-demand (command-line) fuzzy finding of my files, and the power for me to determine exactly how and when my files are uploaded/downloaded.

[jstaf/onedriver](https://github.com/jstaf/onedriver) currently checks the on-demand functionality checkbox that I'm after, and its great. With [`tmux`](https://github.com/tmux/tmux/wiki), I can keep it running in the background (though that's not necessarily the only way to achieve that) with something like:

```bash
$ tmux start-server
$ tmux new-session -d -s onedriver onedriver /onedriver/mount/path/here
```

With [`fd`](https://github.com/sharkdp/fd), I can `$ fd . /onedriver/mount/path/here`  to force `onedriver` to cache the paths of all files that I currently have in OneDrive. It took around 3 minutes for `fd` to find all of my 19,000 files that currently live in my OneDrive and cache their paths. 

`onedriver` checks in with OneDrive every 30 seconds; any changes simply get "appended" to the cached state, which is why every subsequent call to `fd` is almost always instant. If I happen to close `onedriver`, though, I will have to wait around for those 3 minutes to re-obtain that cached state.

Then, I can simply pipe `fd`'s output to [`fzf`](https://github.com/junegunn/fzf) for fuzzy finding with `$ fd . /onedriver/mount/path/here | fzf`, and fuzzy find all of my OneDrive files.

That's progress, but I have only achieved fuzzy finding across the paths of my OneDrive files. Nothing ever gets downloaded or uploaded, though we do have some of the puzzle pieces to achieve that kind of behavior.

`onedriver` generally works in the following manner:

- If you open it, it gets downloaded.
- If you save it, it gets *replaced*.

Emphasis on that second rule. To download the latest version of a file from OneDrive with `onedriver`, you simply have to open it (e.g., `nvim /onedriver/mount/path/here/plus/path/to/file`). But `onedriver` is not a sync client; when you save the file you just downloaded (presumably after making some changes to it), `onedriver` deletes the OneDrive version of that file, and uploads your own local copy. In this process (for some reason), creation timestamps of files get lost. And I don't like that.

However, files downloaded by `onedriver` and uploaded by `onedrive` *do* maintain their creation timestamps. Hence, I propose the following workaround:

1. On-demand, fuzzy-find download stuff with `onedriver` (or just download stuff normally with `$ onedrive --synchronize`)
2. Upload stuff with `$ onedrive --synchronize` strictly, perhaps disposing of command line options like `--single-directory <dir>`, `--upload-only`, and `--no-remote-delete`.

Here's a stupidly simple bash script that does the first part, using `fzf` and `fd`:

```bash
#!/usr/bin/env bash

onedriver_base_path="/path/to/onedriver/mountpoint/here"
onedrive_base_path="/path/to/onedrive/main/sync/folder/here"

fzf_result=`fd . /media/kalyan/od | fzf`

relative_path=`echo $fzf_result | sed "s|"$onedriver_base_path"/||"`

mkdir -p `dirname "$onedrive_base_path$relative_path"`
cp $fzf_result "$onedrive_base_path$relative_path"
```

Make that script available in somewhere in your `$PATH`, and use something like `tmux new-window <script-name-here>` to on-demand, fuzzy-find download your OneDrive files. Great! You could add some error checking to make sure that `onedriver` is actually mounted to `/path/to/onedriver/mountpoint/here` and all that, but that comes later.

For the second part, I suppose I could also write a wrapper script around `$ onedrive --synchronize --single-directory <dir> --upload-only --no-remote-delete` to achieve [a `--single-file`-like bevavior](https://github.com/abraunegg/onedrive/discussions/2750), but that's not an immediate concern for me at the moment.

Anyways, 

> the kind of behavior that I'm \[I was\] after \[was\]: on-demand (command line) fuzzy finding of my files, and the power for me to determine exactly how and when my files are uploaded/downloaded,

and I more or less achieved that. Though some [native on-demand functionality for `onedrive`](https://github.com/abraunegg/onedrive/issues/757), [the ability to `ls` your OneDrive folder with `onedrive`](https://github.com/abraunegg/onedrive/discussions/2749), and [some way to specify files instead of folder to sync](https://github.com/abraunegg/onedrive/discussions/2750) would be nice.

If anyone has any suggestions to this workaround, let me know. This was my first time messing around with `tmux` and bash scripting, so I'm certainly still a beginner and learning.

**Comment by kalyanoliveira** [comment]
Now, I speak directly to you, [abraunegg](https://github.com/abraunegg). I'm not going to ping you because I'm not sure if you'd get annoyed by that, so I guess I'll just have to hope you read this lol.

What kind of functionality from what I showed above do you think would be feasible to implement natively in the OneDrive client that you created? I don't wanna run around raising feature requests like crazy.

**Comment by abraunegg** [comment]
@kalyanoliveira 
> What kind of functionality from what I showed above do you think would be feasible to implement natively in the OneDrive client that you created? I don't wanna run around raising feature requests like crazy.

As per my previous responses - please raise a feature request. I cant be clearer in how that is articulated to you .. 

My focus at the moment is delivery of v2.5.x and the major performance enhancement this brings.

**Reply by kalyanoliveira** [reply]
Will do! Once I get some time to play around with all this again, I'll make the feature requests.

**Comment by abraunegg** [comment]
@kalyanoliveira 
Comments on your other points that you raise

> `onedriver` deletes the OneDrive version of that file, and uploads your own local copy. In this process (for some reason), creation timestamps of files get lost. And I don't like that.

Raise a bug report here: https://github.com/jstaf/onedriver/issues as that is probably a bug. The online file should be replaced with timestamps kept ... but OneDrive Account Types + Graph API do odd things - for example Personal Accounts act totally different to Business Accounts for timestamp keeping, then there is the whole SharePoint debacle where Microsoft will *modify* your file post upload changing not only the timestamp, but adding data to your file - thus your integrity of that file is gone. Please read https://github.com/OneDrive/onedrive-api-docs/issues/935 for details.

`onedriver` currently does not support Shared Folders and SharePoint .. but if you are using a Business Account - all your data is being stored at the back end on SharePoint thus impacted by the API bug.
